frontend/src/providers/AuthProvider.vue

```diff
@@ -0,0 +1,275 @@
+<script setup lang="ts">
+import {
+    ref,
+    provide,
+    onMounted,
+    onBeforeUnmount,
+    reactive,
+    readonly,
+} from "vue";
+import { Auth } from "aws-amplify";
+import {
+    CognitoUser,
+    CognitoAccessToken,
+    CognitoIdToken,
+    CognitoRefreshToken,
+} from "amazon-cognito-identity-js";
+import type { CognitoUserSession } from "amazon-cognito-identity-js";
+
+import type {
+    IdpTypes,
+    AuthContext,
+    FamLoginUser,
+    AuthState,
+} from "@/types/AuthTypes";
+import { AUTH_KEY } from "@/constants/InjectionKeys";
+import { FOUR_MINUTES, HALF_HOUR } from "@/constants/TimeUnits";
+import { IdpProvider } from "@/enum/IdpEnum";
+import { EnvironmentSettings } from "@/services/EnvironmentSettings";
+import { authState } from "@/providers/authState";
+import authLoadingSvg from "@/assets/images/auth-loading.svg";
+
+const environmentSettings = new EnvironmentSettings();
+const REFRESH_INTERVAL = FOUR_MINUTES;
+const INACTIVITY_TIMEOUT = HALF_HOUR;
+
+let refreshIntervalId: number | null = null;
+let inactivityTimeoutId: number | null = null;
+const isLoading = ref(false); // Loading state for animation
+
+/**
+ * Resets the inactivity timeout and sets a new timeout to log the user out after a period of inactivity.
+ * If the user is inactive (no mouse or keyboard input) for the defined `INACTIVITY_TIMEOUT`,
+ * and is still authenticated, they will be logged out and the silent refresh will stop.
+ */
+const resetInactivityTimeout = () => {
+    if (inactivityTimeoutId) clearTimeout(inactivityTimeoutId);
+
+    // Set a timeout that logs the user out after inactivity
+    inactivityTimeoutId = window.setTimeout(() => {
+        if (authState.value.isAuthenticated) {
+            console.log("User inactive, logging out.");
+            logout();
+        }
+    }, INACTIVITY_TIMEOUT);
+};
+
+/**
+ * Stops the silent token refresh process.
+ */
+const stopSilentRefresh = () => {
+    if (refreshIntervalId) clearInterval(refreshIntervalId);
+    refreshIntervalId = null;
+};
+
+/**
+ * Logs in the user using the specified identity provider.
+ * @param {IdpTypes} idP The identity provider type (IDIR, BCEIDBUSINESS).
+ */
+const login = async (idP: IdpTypes) => {
+    try {
+        const customProvider =
+            idP === IdpProvider.IDIR
+                ? environmentSettings.getIdentityProviderIdir()
+                : environmentSettings.getIdentityProviderBceid();
+        await Auth.federatedSignIn({ customProvider });
+    } catch (error) {
+        console.error("Login failed:", error);
+    }
+};
+
+/**
+ * Logs the user out and resets authentication state.
+ */
+const logout = async () => {
+    await Auth.signOut();
+    stopSilentRefresh();
+    authState.value = reactive<AuthState>({
```

Comment:
Something a bit odd, you can directly setting the "authSate.value" with the object you need that I guess vue `ref` will reflect the change, or is it not? why wrapping into another '**reactive**' and not '**ref**'?

Reply:
The reason we wrap the state in reactive is due to the fact that the authState is exposed as readonly to external components through the AuthProvider. Since it’s made readonly, we can’t directly modify it from outside, ensuring that no external mutation happens accidentally. When logging out, we create a new reactive state internally to bypass the readonly restriction and properly update the state. So while Vue does reflect changes through ref, the readonly wrapper necessitates this approach for safe mutation management.

Reply:
Thanks. If after you experiment direct assign object does not work, I only wonder if wrapping in ref() is the suitable one not reactive() (but maybe they have no conflict..).